binary_search_tree/tree.py

```diff
@@ -17,32 +17,121 @@ def __init__(self):
     # Time Complexity: 
     # Space Complexity: 
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+            return
+            
+        def add_helper(current):
+            if key < current.key:
+                if current.left is not None:
+                    add_helper(current.left)
+                else:
+                    current.left = TreeNode(key, value)
+            else:
+                if current.right is not None:
+                    add_helper(current.right)
+                else:
+                    current.right = TreeNode(key,value)
+        
+        add_helper(self.root)
 
     # Time Complexity: 
     # Space Complexity: 
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+
+        def find_helper(current):
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                if current.left is not None:
+                    return find_helper(current.left)
+            else:
+                if current.right is not None:
+                    return find_helper(current.right)
+        return find_helper(self.root)
 
     # Time Complexity: 
     # Space Complexity: 
     def inorder(self):
-        pass
+        return_list = []
+        current = self.root
+        if self.root == None:
+            return return_list
+
+        def in_order_helper(current):
+            if current.left is not None:
+                in_order_helper(current.left)
+            return_list.append({'key': current.key, 'value': current.value})
+            if current.right is not None:
+                in_order_helper(current.right)
+        
+        in_order_helper(current)
+        return return_list
```

Comment:
This works too, but could be a little cleaner. (But nothing so bad that I think it's worth docking points!)

For example, the `if` block on line 60 could come first before the `current` variable is initialized. You're not going to need it if there's nothing in the tree, so you can skip initialization.

Additionally the definition of `current` outside of `in_order_helper` isn't really necessary in this case as it will not change value before it's passed in on line 70, so you could just pass in `self.root` there.

I had long thing written up about `return_list` in the same way, but as I thought about it, while usually writing nested functions like `in_order_helper` that reach outside of their scope is not preferred, doing it in this case does make it a bit cleaner as `in_order_helper` doesn't need to return anything itself.

Just be aware that this type of "closure capture" is often something that can lead to bugs, especially if you begin working with asynchronous programming where you might have multiple threads working together to solve a problem. (But that's a really advanced subject!)

Reply:
The same comments apply to `preorder` and `postorder`, so I won't repeat them.